cf-agent/audit_mode.c

```diff
@@ -0,0 +1,337 @@
+    struct group *group = getgrgid(st->st_gid);
+
+    /* Inspired by the output from the 'stat' command */
+    char mode_str[10] = {
+        st->st_mode & S_IRUSR ? 'r' : '-',
+        st->st_mode & S_IWUSR ? 'w' : '-',
+        (st->st_mode & S_ISUID
+         ? (st->st_mode & S_IXUSR ? 's' : 'S')
+         : (st->st_mode & S_IXUSR ? 'x' : '-')),
+        st->st_mode & S_IRGRP ? 'r' : '-',
+        st->st_mode & S_IWGRP ? 'w' : '-',
+        (st->st_mode & S_ISGID
+         ? (st->st_mode & S_IXGRP ? 's' : 'S')
+         : (st->st_mode & S_IXGRP ? 'x' : '-')),
+        st->st_mode & S_IROTH ? 'r' : '-',
+        st->st_mode & S_IWOTH ? 'w' : '-',
+        (st->st_mode & S_ISVTX
+         ? (st->st_mode & S_IXOTH ? 't' : 'T')
+         : (st->st_mode & S_IXOTH ? 'x' : '-')),
+        '\0'
+    };
+    printf("Size: %ju\n", (uintmax_t) st->st_size);
+    printf("Access: (%04o/%s)  ", st->st_mode & CHMOD_MODE_BITS, mode_str);
+    printf("Uid: (%ju/%s)   ",
+           (uintmax_t)st->st_uid, owner->pw_name);
+    printf("Gid: (%ju/%s)\n",
+           (uintmax_t)st->st_gid, group->gr_name);
+
+#define MAX_TIMESTAMP_LENGTH (sizeof("2020-10-05 12:56:18 +0200"))
+    char buf[MAX_TIMESTAMP_LENGTH] = {0};
+
+    size_t ret = strftime(buf, sizeof(buf), "%Y-%m-%d %H:%M:%S %z",
+                          localtime((time_t*) &(st->st_atime)));
+    assert(ret <= MAX_TIMESTAMP_LENGTH);
+    printf("Access: %s\n", buf);
+
+    ret = strftime(buf, sizeof(buf), "%Y-%m-%d %H:%M:%S %z",
+                   localtime((time_t*) &(st->st_mtime)));
+    assert(ret <= MAX_TIMESTAMP_LENGTH);
+    printf("Modify: %s\n", buf);
+
+    ret = strftime(buf, sizeof(buf), "%Y-%m-%d %H:%M:%S %z",
+                   localtime((time_t*) &(st->st_ctime)));
+    assert(ret <= MAX_TIMESTAMP_LENGTH);
+    printf("Change: %s\n", buf);
+}
+#else  /* !__MINGW32__ */
+static void ManifestStatInfo(const struct stat *st)
+{
+    assert(st != NULL);
+
+    /* Inspired by the output from the 'stat' command */
+    char mode_str[10] = {
+        st->st_mode & S_IRUSR ? 'r' : '-',
+        st->st_mode & S_IWUSR ? 'w' : '-',
+        st->st_mode & S_IXUSR ? 'x' : '-',
+        st->st_mode & S_IRGRP ? 'r' : '-',
+        st->st_mode & S_IWGRP ? 'w' : '-',
+        st->st_mode & S_IXGRP ? 'x' : '-',
+        st->st_mode & S_IROTH ? 'r' : '-',
+        st->st_mode & S_IWOTH ? 'w' : '-',
+        st->st_mode & S_IXOTH ? 'x' : '-',
+        '\0'
+    };
+    printf("Size: %ju\n", (uintmax_t) st->st_size);
+    printf("Access: (%04o/%s)  ", st->st_mode, mode_str);
+    printf("Uid: %ju   ", (uintmax_t)st->st_uid);
+    printf("Gid: %ju\n", (uintmax_t)st->st_gid);
+
+#define MAX_TIMESTAMP_SIZE (sizeof("2020-10-05 12:56:18 +0200"))
+    char buf[MAX_TIMESTAMP_SIZE] = {0};
+
+    size_t ret = strftime(buf, sizeof(buf), "%Y-%m-%d %H:%M:%S %z",
+                          localtime((time_t*) &(st->st_atime)));
+    assert((ret > 0) && (ret <= MAX_TIMESTAMP_SIZE));
```

Comment:
You didn't follow my suggestion exactly. I think it should be less than, not less than or equal, since the return value does not count null byte, while the size constant does.